File: algs/quick_sort.c
//采用分治法进行快速排序
#include<stdio.h>
#include<stdlib.h>

int partion(int a[],int lo,int hi); //找出数组a[lo,hi]的划分点

void quick_sort(int a[],int lo,int hi)
{
  if(lo < hi){               //至少有两个元素才进行划分
    int p = partion(a,lo,hi);
    quick_sort(a,lo,p - 1);
    quick_sort(a,p + 1,hi);
  }
}

int partion(int a[],int lo,int hi)
{
  int x = a[hi];
  int i = lo - 1;
  void exchange(int a[],int i,int hi);
  for(int j = lo;j < hi;++j)
    if(a[j] <= x)
      exchange(a,++i,j);
  ++i;
  exchange(a,i,hi);
  return i;
}

void exchange(int a[],int lo,int hi)
{
  int tmp = a[lo];
  a[lo] = a[hi];
  a[hi] = tmp;
}

int main()
{
  int a[100]={0};
  for(int i = 0;i < 100;++i)
    a[i] = rand() % 100;
  quick_sort(a,0,99);
  for(int i = 0;i < 99;++i){
    printf("%d\t",a[i]);
    if((i + 1) % 10 == 0)
      printf("\n");
    }
  printf("\n");
  return 0;
}

File: algs/quick_sort.py
import random

#快速排序的python实现

def quick_sort(l):
    return quick_sort_helper(l,0,len(l) - 1)

def quick_sort_helper(l,left,right):
    if right > left:
        pos = partition(l,left,right)
        quick_sort_helper(l,left,pos-1)
        quick_sort_helper(l,pos+1,right)

#循环不变式，l[left:pos+1]为小于l[right]的元素,
#           l[pos+1:i] 为大于等于l[right]的元素
#           l[i:]为将要比较的元素
def partition(l,left,right):
    #随机选择一个元素进行划分
    rad = random.randint(left,right)
    exchange(l,rad,right)

    last = l[right]
    pos = left - 1
    i = 0

    for i in range(left,right+1):
        if l[i] < last:
            pos += 1
            exchange(l,pos,i)

    pos += 1
    exchange(l,pos,i)

    return pos

def exchange(l,i,j):
    tmp = l[i]
    l[i] = l[j]
    l[j] = tmp
